### PR TITLE
TEST/EXAMPLES: Add UCP hello world example

### PR DESCRIPTION
**Pull request by ilyanelken**
Signed-off-by: Ilya Nelkenbaum <ilyan@mellanox.com>

### CODE REVIEW

**Comment by mellanox-github** [comment]
Can one of the admins verify this patch?

**Comment by yosefe** [comment]
ok to test

**Comment by yosefe** [comment]
@alinask pls review

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1257/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1872/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1873/ for details.

File: test/examples/ucp_hello_world.c [hunk 1]
```diff
@@ -1,17 +1,487 @@
...
+	struct ucx_context *context = (struct ucx_context *) request;
+
+	context->completed = 1;
+
+	printf("[0x%x] receive handler called with status %d (length %u)\n",
+	       (unsigned int)pthread_self(), status, info->length);
+}
+
+static void wait(ucp_worker_h *ucp_worker, struct ucx_context *context) {
+	while (context->completed == 0)
```
Comment (alinask) [inline test/examples/ucp_hello_world.c]: if you are passing the request from all the wait calls, you can use ucp_request_is_completed

Comment (ilyanelken) [inline test/examples/ucp_hello_world.c]: correct.
The reason is that I want to demonstrate usage of user defined request context.

File: test/examples/ucp_hello_world.c [hunk 2]
```diff
@@ -1,17 +1,487 @@
...
+	status = ucp_worker_create(ucp_context, UCS_THREAD_MODE_SINGLE,
+	                           &ucp_worker);
+	if (status != UCS_OK)
+		goto err_cleanup;
+
+	status = ucp_worker_get_address(ucp_worker, &local_addr, &local_addr_len);
+	if (status != UCS_OK)
+		goto err_worker;
+
+	tls = getenv("UCX_TLS");
```
Comment (alinask) [inline test/examples/ucp_hello_world.c]: you can use the tls from the config

Comment (ilyanelken) [inline test/examples/ucp_hello_world.c]: Sounds good, thanks.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1259/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1875/ for details.

**Comment by yosefe** [comment]
@ilyanelken pls comply to UCX [code style](https://github.com/openucx/ucx/blob/master/doc/CodeStyle)

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1260/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1876/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1263/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1877/ for details.

**Comment by shamisp** [comment]
I would suggest to rebase, otherwise 👍

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1265/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1879/ for details.

File: test/examples/ucp_hello_world.c [hunk 3]
```diff
@@ -1,17 +1,490 @@
...
+#include <sys/socket.h>
+#include <sys/types.h>
+#include <netinet/in.h>
+#include <assert.h>
+#include <netdb.h>
 #include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#define UCX_MAX_ADDR_LEN    256
```
Comment (yosefe) [inline test/examples/ucp_hello_world.c]: this is an incorrect assumption

File: test/examples/ucp_hello_world.c [hunk 4]
```diff
@@ -1,17 +1,490 @@
...
+    ucp_tag_message_h msg_tag;
+    ucs_status_t status;
+    ucp_ep_h server_ep;
+    struct msg *msg;
+    struct ucx_context *request;
+    int ret = -1;
+
+    /* Send client UCX address to server */
+    status = ucp_ep_create(*ucp_worker, peer_addr, &server_ep);
+    if (status != UCS_OK)
```
Comment (yosefe) [inline test/examples/ucp_hello_world.c]: code style

File: test/examples/ucp_hello_world.c [hunk 5]
```diff
@@ -1,17 +1,490 @@
...
+    } else if (UCS_PTR_STATUS(request) != UCS_OK) {
+        printf("UCX address message was scheduled for send\n");
+        wait(ucp_worker, request);
+        ucp_request_release(request);
+    }
+
+    free (msg);
+
+    /* Receive test string from server */
+    for (;;) {
```
Comment (yosefe) [inline test/examples/ucp_hello_world.c]: need to call ucp_worker_progress in the loop

File: test/examples/ucp_hello_world.c [hunk 6]
```diff
@@ -1,17 +1,490 @@
...
+    ucp_config_t *config;
+    ucs_status_t status;
+
+    /* UCP handler objects */
+    ucp_context_h ucp_context;
+    ucp_worker_h ucp_worker;
+
+    /* OOB connection vars */
+    char recv_buf[64];
+    size_t recv_size = sizeof(recv_buf);
```
Comment (yosefe) [inline test/examples/ucp_hello_world.c]: unused variable

File: test/examples/ucp_hello_world.c [hunk 7]
```diff
@@ -1,17 +1,490 @@
...
+    printf("[0x%x] local address length: %u\n",
+           (unsigned int)pthread_self(), local_addr_len);
+
+    if (local_addr_len > UCX_MAX_ADDR_LEN) {
+        fprintf(stderr, "local UCX address is larger than %d\n",
+            UCX_MAX_ADDR_LEN);
+        goto err_addr;
+    }
+
+    /* Assume peer have the same UCX address length */
```
Comment (yosefe) [inline test/examples/ucp_hello_world.c]: this is an incorrect assumption

File: test/examples/ucp_hello_world.c [hunk 8]
```diff
@@ -1,17 +1,490 @@
...
+    ucp_worker_release_address(ucp_worker, local_addr);
+
+err_worker:
+    ucp_worker_destroy(ucp_worker);
+
+err_cleanup:
+    ucp_cleanup(ucp_context);
+
+err:
+    fflush(stderr);
```
Comment (yosefe) [inline test/examples/ucp_hello_world.c]: no need to flush here

**Comment by yosefe** [comment]
we need to add to jenkins running this example code

**Comment by MattBBaker** [comment]
ORNL failure this time was an ICE, no PR related.

File: test/examples/ucp_hello_world.c (hunk 7 above)
Comment (shamisp) [inline test/examples/ucp_hello_world.c]: good point...

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1890/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1275/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1276/ for details (Mellanox internal link).

**Comment by ilyanelken** [comment]
bot:retest

File: contrib/test_jenkins.sh [hunk 9]
```diff
@@ -99,6 +99,21 @@ if [ -n "$JENKINS_RUN_TESTS" ]; then
...
     echo "LD_LIBRARY_PATH=${LD_LIBRARY_PATH}"
 
+    # compile and run UCP hello world example
+    gcc -o ./ucp_hello_world $ucx_inst/share/ucx/examples/ucp_hello_world.c -lucp -I${ucx_inst}/include -L${ucx_inst}/lib
+
+    echo Running UCP hello world server
+    ./ucp_hello_world &
+    hw_pid=$!
+
+    echo Running UCP hello world client
```
Comment (yosefe) [inline contrib/test_jenkins.sh]: might want to add small sleep here to let server create socket

Comment (ilyanelken) [inline contrib/test_jenkins.sh]: sounds good, I will add that

File: contrib/test_jenkins.sh [hunk 10]
```diff
@@ -99,6 +99,21 @@ if [ -n "$JENKINS_RUN_TESTS" ]; then
...
+    ./ucp_hello_world $(hostname)
+
+    # make sure server process in not running
+    wait $hw_pid
+
+    rm -f ./ucp_hello_world
```
Comment (alinask) [inline contrib/test_jenkins.sh]: why are you removing it?

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1891/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1892/ for details.

File: contrib/test_jenkins.sh (hunk 10 above)
Comment (ilyanelken) [inline contrib/test_jenkins.sh]: because we don't need it (same is done for "active_message" example)

Comment (alinask) [inline contrib/test_jenkins.sh]: i thought we may want to keep it there in case jenkins fails and we want to reproduce...

Comment (ilyanelken) [inline contrib/test_jenkins.sh]: we have bash "set -e" option, which causes script to exit on each failure.
So if we will have failure before "rm -f ...."  the script will exit and the executable will be safe
from removal.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1893/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1277/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1894/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1278/ for details (Mellanox internal link).

**Comment by ilyanelken** [comment]
bot:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1895/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1279/ for details (Mellanox internal link).

**Comment by ilyanelken** [comment]
@alinask @yosefe please have a look

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1896/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1897/ for details.

File: test/examples/ucp_hello_world.c [hunk 11]
```diff
@@ -1,17 +1,508 @@
...
+                    UCS_PTR_STATUS(request));
+            free(msg);
+            goto err;
+        } else {
+            wait(ucp_worker, request);
+            ucp_request_release(request);
+            printf("UCX address message was received\n");
+        }
+
+        break;
```
Comment (yosefe) [inline test/examples/ucp_hello_world.c]: looks like msg is not released here?

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1903/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1283/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1906/ for details.

**Comment by alinask** [comment]
:+1: 

@ilyanelken can you please squash the commits?

**Comment by ilyanelken** [comment]
sure

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1909/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1289/ for details (Mellanox internal link).

**Pull request merged**